File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.js
function solution(board, moves) {
    let answer = 0;
    const stk = [];
    const n = board[0].length;
    
    const getNumberInBoard = (m) => {
        for(let i=0; i<n; i++){
            const v = board[i][m-1]
            if(v){
                board[i][m-1] = 0;
                return v;
            }
        }
        return 0;
    }
    
    moves.forEach(m => {
        const v = getNumberInBoard(m);
        
        if(v !== 0){
            if(stk.length !== 0 && stk[stk.length -1] === v){
                stk.pop();
                answer += 2;
            }else {
                stk.push(v);
            }    
        }
    })
    return answer;
}

File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.cpp
#include <bits/stdc++.h>
using namespace std;

int solution(vector<vector<int>> board, vector<int> moves) {
    int ret = 0;
    int n = board[0].size();
    
    vector<stack<int>> stks(n);
    stack<int> stk;
    
    for (int i=n-1; i>=0; i--){
        for(int j=0; j<n; j++){
            if(board[i][j] != 0){
                stks[j].push(board[i][j]);    
            }
        }
    }
    
    for (int m : moves){
        if(stks[m-1].empty()) continue;
        
        int here = stks[m-1].top();
        cout << here << '\n';
        stks[m-1].pop();
        
        if(!stk.empty() && stk.top() == here){
            stk.pop();
            ret += 2;
        }else{
            stk.push(here);
        }
    }
    
    return ret;
}